OSX: add homebrew lib directory to library paths on OSX

buggins opened the pull request:
OSX Sonoma arm64/m3 bindbc cannot find libfreetype installed with homebrew unless library path variable is set
```DYLD_LIBRARY_PATH=/opt/homebrew/lib```

Please let me know if there is some better way for adding the path.
Probably, better place is some common bindbc package?

Commit:
OSX: add homebrew lib directory to library paths on OSX

ichordev commented:
[This line](https://github.com/BindBC/bindbc-loader/blob/af7667b39bb31b841a5dc07fff6231f0df95acb5/source/bindbc/loader/codegen.d#L31) in BindBC-Loader *should* be doing the same thing. Is FreeType's shared library not named `libFreeType.dylib` under Homebrew?

Also, have you tested this commit? From reading my own code, it seems I made an error in handling those paths that means that they shouldn't work. If they are working, then that is weird indeed.

buggins commented:
I'm trying to get dlangui working out of the box on my MacBook Pro M3 laptop.
Latest OSX Sonoma, latest homebrew, sdl and freetype are installed with brew.

dlangui has bindbc freetype, sdl and opengl dependencies.
All of them were set to "~>1.0.0". I've upgraded them to latest versions.
```
  Up-to-date bindbc-loader 1.1.4: target for configuration [noBC] is up to date.
  Up-to-date bindbc-freetype 1.1.1: target for configuration [dynamic] is up to date.
  Up-to-date bindbc-opengl 1.1.0: target for configuration [dynamic] is up to date.
  Up-to-date bindbc-common 0.1.2: target for configuration [noBC] is up to date.
  Up-to-date bindbc-sdl 1.4.5: target for configuration [dynamic] is up to date.
``` 
But it did not help. Still, app which uses dlangui cannot find freetype and sdl w/o workaround DYLD_LIBRARY_PATH=/opt/homebrew/lib

```
% ls -al /opt/homebrew/lib | grep freetype
lrwxr-xr-x    1 vlopatin  admin    49 Feb 26 13:37 libfreetype.6.dylib -> ../Cellar/freetype/2.13.2/lib/libfreetype.6.dylib
lrwxr-xr-x    1 vlopatin  admin    43 Feb 26 13:37 libfreetype.a -> ../Cellar/freetype/2.13.2/lib/libfreetype.a
lrwxr-xr-x    1 vlopatin  admin    47 Feb 26 13:37 libfreetype.dylib -> ../Cellar/freetype/2.13.2/lib/libfreetype.dylib

% ls -al /opt/homebrew/lib | grep sdl     
lrwxr-xr-x    1 vlopatin  admin    45 Feb 26 14:08 libSDL2-2.0.0.dylib -> ../Cellar/sdl2/2.30.0/lib/libSDL2-2.0.0.dylib
lrwxr-xr-x    1 vlopatin  admin    35 Feb 26 14:08 libSDL2.a -> ../Cellar/sdl2/2.30.0/lib/libSDL2.a
lrwxr-xr-x    1 vlopatin  admin    39 Feb 26 14:08 libSDL2.dylib -> ../Cellar/sdl2/2.30.0/lib/libSDL2.dylib
lrwxr-xr-x    1 vlopatin  admin    40 Feb 26 14:08 libSDL2_test.a -> ../Cellar/sdl2/2.30.0/lib/libSDL2_test.a
lrwxr-xr-x    1 vlopatin  admin    39 Feb 26 14:08 libSDL2main.a -> ../Cellar/sdl2/2.30.0/lib/libSDL2main.a
```
Trying to test the fix.
Added directory with cloned bindbc-freetype to dub.selections.json
```
    Building bindbc-freetype ~master: building configuration [dynamic]
../../../bindbc-freetype/source/bindbc/freetype/dynload.d-mixin-74(74,7): Error: undefined identifier `bindModuleSymbols` in module `ft.bdf`
Error ldc2 failed with exit code 1.
```

Tried to build just a bindbc-freetype library. Build is ok for configurations **static** and **staticBC**, fails with the same error for configurations **dynamic**, **dynamicBC**
Upgrade to latest bindbc-common and bindbc-loader does not help.

Tried to build older versions.

Commit / failed module
31ce5a66a37685bd468160732bdd846dbfeb4a21 - ft.bdf
9f11d3b641f24e3918cadb210f0f2ab21bc72a83 - ft.bdf
c61617d1edccaff6e22d475b1b59e57fd67602e5 - ft.bdf
914dd1aff1548dbbbb9c64b93208e903d8864f4c - ft.bdf ft.logging
6998d302a24dfe0428fe94331737b98bef3e301a - ft.bdf ft.logging
704b7425734a4b4419350ea7361a506de4bbdc29 - ft.bdf ft.logging
091cb9a6267c749d3bc2773227aeb0fdb704d0c4 — build is ok

example:
```
% git checkout 914dd1aff1548dbbbb9c64b93208e903d8864f4c
% dub build --config=dynamic  
    Starting Performing "debug" build using ldc2 for aarch64, arm_hardfloat.
    Building bindbc-freetype ~master: building configuration [dynamic]
source/bindbc/freetype/dynload.d-mixin-74(74,7): Error: undefined identifier `bindModuleSymbols` in module `ft.bdf`
source/bindbc/freetype/dynload.d-mixin-74(74,11): Error: undefined identifier `bindModuleSymbols` in module `ft.logging`
Error ldc2 failed with exit code 1.
```

buggins commented:
Build error reason for master version is version(linux) in ft.bdf module:
```
diff --git a/source/ft/bdf.d b/source/ft/bdf.d
index 38ef517..b8419f0 100644
--- a/source/ft/bdf.d
+++ b/source/ft/bdf.d
@@ -7,7 +7,7 @@
 +/
 module ft.bdf;
 
-version(linux):
+//version(linux):
```

If commented out, fails on 2 import symbols - probably, from ft.bdf

I've added debug logs to loadFreeType:

```
loadFreetype(/usr/X11/lib/)
     failed to load library /usr/X11/lib/
loadFreetype(/opt/X11/lib/)
     failed to load library /opt/X11/lib/
loadFreetype(/opt/homebrew/lib/)
     failed to load library /opt/homebrew/lib/
loadFreetype(libfreetype.6.dylib)
     failed to load library libfreetype.6.dylib
loadFreetype(/opt/homebrew/lib/libfreetype.6.dylib)
loadFreetype(/opt/homebrew/lib/libfreetype.6.dylib) -- library is loaded, checking symbols
loadFreetype(/opt/homebrew/lib/libfreetype.6.dylib) -- symbol import errors: 2
loadFreetype(freetype.6)
     failed to load library freetype.6
loadFreetype(/Library/Frameworks/freetype.6.framework/freetype.6)
     failed to load library /Library/Frameworks/freetype.6.framework/freetype.6
loadFreetype(/System/Library/Frameworks/freetype.6.framework/freetype.6)
     failed to load library /System/Library/Frameworks/freetype.6.framework/freetype.6
loadFreetype(libfreetype.dylib)
     failed to load library libfreetype.dylib
loadFreetype(/opt/homebrew/lib/libfreetype.dylib)
loadFreetype(/opt/homebrew/lib/libfreetype.dylib) -- library is loaded, checking symbols
loadFreetype(/opt/homebrew/lib/libfreetype.dylib) -- symbol import errors: 2
loadFreetype(freetype)
     failed to load library freetype
loadFreetype(/Library/Frameworks/freetype.framework/freetype)
     failed to load library /Library/Frameworks/freetype.framework/freetype
loadFreetype(/System/Library/Frameworks/freetype.framework/freetype)
     failed to load library /System/Library/Frameworks/freetype.framework/freetype
```

Commit:
Fix build errors. OpenType Validation is disabled - probably version check is needed?

buggins commented:
Fixed build error caused by version(linux) in ft.bdf . Removed my previous change - it does not take any effect.
Symbols import for ft.otval module are disabled now - at least not found for my freetype version.
Probably they should be put under version check.

List of library names to load looks very strange (see my previous message) - starts with directory names w/o file names.

ichordev commented:
> List of library names to load looks very strange (see my previous message) - starts with directory names w/o file names.

Oops yes, this is definitely the result of that error in my code that I noticed. It’s an issue with BindBC-Loader’s codegen.d, not this library. I’ll patch it tonight when I can.

ichordev commented:
> Build error reason for master version is version(linux) in ft.bdf module:
> 
> ```
> diff --git a/source/ft/bdf.d b/source/ft/bdf.d
> index 38ef517..b8419f0 100644
> --- a/source/ft/bdf.d
> +++ b/source/ft/bdf.d
> @@ -7,7 +7,7 @@
>  +/
>  module ft.bdf;
>  
> -version(linux):
> +//version(linux):
> ```
> 
> If commented out, fails on 2 import symbols - probably, from ft.bdf
> 
> I've added debug logs to loadFreeType:
> 
> ```
> loadFreetype(/usr/X11/lib/)
>      failed to load library /usr/X11/lib/
> loadFreetype(/opt/X11/lib/)
>      failed to load library /opt/X11/lib/
> loadFreetype(/opt/homebrew/lib/)
>      failed to load library /opt/homebrew/lib/
> loadFreetype(libfreetype.6.dylib)
>      failed to load library libfreetype.6.dylib
> loadFreetype(/opt/homebrew/lib/libfreetype.6.dylib)
> loadFreetype(/opt/homebrew/lib/libfreetype.6.dylib) -- library is loaded, checking symbols
> loadFreetype(/opt/homebrew/lib/libfreetype.6.dylib) -- symbol import errors: 2
> loadFreetype(freetype.6)
>      failed to load library freetype.6
> loadFreetype(/Library/Frameworks/freetype.6.framework/freetype.6)
>      failed to load library /Library/Frameworks/freetype.6.framework/freetype.6
> loadFreetype(/System/Library/Frameworks/freetype.6.framework/freetype.6)
>      failed to load library /System/Library/Frameworks/freetype.6.framework/freetype.6
> loadFreetype(libfreetype.dylib)
>      failed to load library libfreetype.dylib
> loadFreetype(/opt/homebrew/lib/libfreetype.dylib)
> loadFreetype(/opt/homebrew/lib/libfreetype.dylib) -- library is loaded, checking symbols
> loadFreetype(/opt/homebrew/lib/libfreetype.dylib) -- symbol import errors: 2
> loadFreetype(freetype)
>      failed to load library freetype
> loadFreetype(/Library/Frameworks/freetype.framework/freetype)
>      failed to load library /Library/Frameworks/freetype.framework/freetype
> loadFreetype(/System/Library/Frameworks/freetype.framework/freetype)
>      failed to load library /System/Library/Frameworks/freetype.framework/freetype
> ```

That’s not how you’re meant to make modules optional though. The version statement just shouldn’t have included the joinFnBinds mixin. I’ll fix this along with the BindBC-Loader issue… stupid oversight, I must’ve forgotten to test for that on macOS.

ichordev commented:
Sorry for the oversight... I discovered that BindBC-FreeType wasn't updated to v1.2.0 in dub, so you've been on v1.1.1 this whole time. I've just released v1.2.1 (and manually updated it in dub...). The problem with `bdf.d` should be fixed. I couldn't see anything wrong with `logger.d`, so maybe it was already fixed in v1.2.0?

Try again with v1.2.1 and see if it's not loading `/opt/homebrew/lib/libfreetype.6.dylib`. It should be.

buggins commented:
setting version to v1.2.1 does not help.
Tried DYLD_LIBRARY_PATH=/opt/homebrew/lib setting - still not working - probably, due to **ft.otval**

ichordev commented:
Ah I must've missed the mention of that module. Its functions were erroneously prefixed with `da_` for some reason. That issue goes all the way back to BindBC-FreeType's original commit, 5 years ago.

Before my restructuring of the library, the issue would've only manifested if someone had tried using those functions with the static bindings. (It would've produced a linker error)

I have fixed the symbol names in `ft/otval.d` in v1.2.2.

buggins commented:
v1.2.2 seems working.
Thank you!

buggins commented:
Issue is fixed. Closing PR.
